File: src/react-native/android/src/main/java/ai/ansight/reactnative/AnsightBridgeModule.kt
package ai.ansight.reactnative

import ai.ansight.runtime.AnsightEventType
import ai.ansight.runtime.AnsightOptions
import ai.ansight.runtime.AnsightRuntime
import ai.ansight.runtime.AnsightToolDescriptor
import ai.ansight.runtime.PairingOpenOptions
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap

class AnsightBridgeModule(
    reactContext: ReactApplicationContext,
) : ReactContextBaseJavaModule(reactContext) {
    override fun getName(): String = "AnsightBridgeModule"

    @ReactMethod
    fun initialize(options: ReadableMap?, promise: Promise) {
        runCatching {
            val application = reactApplicationContext.applicationContext
            require(application is android.app.Application) {
                "React application context must be backed by an Application instance."
            }

            AnsightRuntime.initialize(application, options.toOptions())
            promise.resolve(null)
        }.onFailure { promise.reject("ansight_initialize_failed", it) }
    }

    @ReactMethod
    fun activate(promise: Promise) {
        runCatching {
            AnsightRuntime.activate()
            promise.resolve(null)
        }.onFailure { promise.reject("ansight_activate_failed", it) }
    }

    @ReactMethod
    fun deactivate(promise: Promise) {
        AnsightRuntime.deactivate()
        promise.resolve(null)
    }

    @ReactMethod
    fun clear(promise: Promise) {
        AnsightRuntime.clear()
        promise.resolve(null)
    }

    @ReactMethod
    fun metric(value: String, channel: Double?, promise: Promise) {
        runCatching {
            AnsightRuntime.metric(value.toLong(), (channel ?: 255.0).toInt())
            promise.resolve(null)
        }.onFailure { promise.reject("ansight_metric_failed", it) }
    }

    @ReactMethod
    fun event(label: String, options: ReadableMap?, promise: Promise) {
        runCatching {
            AnsightRuntime.event(
                label = label,
                type = options?.getString("type")?.let(AnsightEventType::valueOf) ?: AnsightEventType.Info,
                details = options?.getString("details"),
                channel = options?.getDouble("channel")?.toInt() ?: 255,
                id = options?.getString("id") ?: java.util.UUID.randomUUID().toString(),
            )
            promise.resolve(null)
        }.onFailure { promise.reject("ansight_event_failed", it) }
    }

    @ReactMethod
    fun openSession(pairingJson: String, options: ReadableMap, promise: Promise) {
        runCatching {
            val result = AnsightRuntime.openSession(
                pairingJson = pairingJson,
                options = PairingOpenOptions(
                    clientName = options.getString("clientName").orEmpty(),
                    manualHostAddress = options.getString("manualHostAddress").orEmpty(),
                    expectedAppId = options.getString("expectedAppId"),
                    profileOverride = options.getMap("profileOverride").toStringMap(),
                ),
            )

            val map = Arguments.createMap().apply {
                putBoolean("success", result.success)
                putString("message", result.message)
                putString("sessionId", result.sessionId)
            }
            promise.resolve(map)
        }.onFailure { promise.reject("ansight_open_session_failed", it) }
    }

    @ReactMethod
    fun completeSession(promise: Promise) {
        AnsightRuntime.completeSession()
        promise.resolve(null)
    }

    @ReactMethod
    fun closeSession(promise: Promise) {
        AnsightRuntime.closeSession()
        promise.resolve(null)
    }

    @ReactMethod
    fun registerTool(tool: ReadableMap, promise: Promise) {
        runCatching {
            AnsightRuntime.registerTool(
                AnsightToolDescriptor(
                    id = tool.getString("id").orEmpty(),
                    name = tool.getString("name").orEmpty(),
                    scope = tool.getString("scope") ?: "Read",
                ),
            )
            promise.resolve(null)
        }.onFailure { promise.reject("ansight_register_tool_failed", it) }
    }

    @ReactMethod
    fun getDebugSnapshot(promise: Promise) {
        val snapshot = AnsightRuntime.snapshot()
        val map = Arguments.createMap().apply {
            putBoolean("initialized", snapshot.initialized)
            putBoolean("active", snapshot.active)
            putBoolean("sessionOpen", snapshot.sessionOpen)
            putInt("metricsRecorded", snapshot.metricsRecorded)
            putInt("eventsRecorded", snapshot.eventsRecorded)
            putInt("registeredTools", snapshot.registeredTools)
            putString("sessionMessage", snapshot.sessionMessage)
            snapshot.lastMetric?.let {
                putMap("lastMetric", Arguments.createMap().apply {
                    putDouble("value", it.value.toDouble())
                    putInt("channel", it.channel)
                    putDouble("capturedAtEpochMs", it.capturedAtEpochMs.toDouble())
                })
            }
            snapshot.lastEvent?.let {
                putMap("lastEvent", Arguments.createMap().apply {
                    putString("id", it.id)
                    putString("label", it.label)
                    putString("type", it.type.name)
                    putString("details", it.details)
                    putInt("channel", it.channel)
                    putDouble("capturedAtEpochMs", it.capturedAtEpochMs.toDouble())
                })
            }
        }
        promise.resolve(map)
    }

    private fun ReadableMap?.toOptions(): AnsightOptions {
        if (this == null) {
            return AnsightOptions()
        }

        return AnsightOptions(
            sampleFrequencyMilliseconds = getIntOrDefault("sampleFrequencyMilliseconds", 500),
            retentionPeriodSeconds = getIntOrDefault("retentionPeriodSeconds", 600),
            enableFramesPerSecond = if (hasKey("enableFramesPerSecond")) getBoolean("enableFramesPerSecond") else true,
            additionalChannels = getArray("additionalChannels").toChannels(),
        )
    }

    private fun ReadableMap.getIntOrDefault(key: String, defaultValue: Int): Int {
        return if (hasKey(key)) getDouble(key).toInt() else defaultValue
    }

    private fun ReadableArray?.toChannels(): List<ai.ansight.runtime.AnsightChannel> {
        if (this == null) {
            return emptyList()
        }

        return (0 until size()).mapNotNull { index ->
            getMap(index)?.let {
                ai.ansight.runtime.AnsightChannel(
                    id = it.getDouble("id").toInt(),
                    name = it.getString("name").orEmpty(),
                    colorHex = it.getString("colorHex"),
                )
            }
        }
    }

    private fun ReadableMap?.toStringMap(): Map<String, String> {
        if (this == null) {
            return emptyMap()
        }

        val map = linkedMapOf<String, String>()
        val iterator = keySetIterator()
        while (iterator.hasNextKey()) {
            val key = iterator.nextKey()
            map[key] = getDynamic(key).asString()
        }
        return map
    }
}

File: src/react-native/ios/AnsightBridgeModule.swift
import AnsightKit
import Foundation
import React

@objc(AnsightBridgeModule)
final class AnsightBridgeModule: NSObject {
    @objc
    static func requiresMainQueueSetup() -> Bool {
        false
    }

    @objc(initialize:resolver:rejecter:)
    func initialize(
        _ options: NSDictionary?,
        resolver resolve: RCTPromiseResolveBlock,
        rejecter reject: RCTPromiseRejectBlock
    ) {
        do {
            try AnsightRuntime.shared.initialize(options: options.toOptions())
            resolve(nil)
        } catch {
            reject("ansight_initialize_failed", error.localizedDescription, error)
        }
    }

    @objc(activate:rejecter:)
    func activate(
        _ resolve: RCTPromiseResolveBlock,
        rejecter reject: RCTPromiseRejectBlock
    ) {
        do {
            try AnsightRuntime.shared.activate()
            resolve(nil)
        } catch {
            reject("ansight_activate_failed", error.localizedDescription, error)
        }
    }

    @objc(deactivate:rejecter:)
    func deactivate(
        _ resolve: RCTPromiseResolveBlock,
        rejecter reject: RCTPromiseRejectBlock
    ) {
        AnsightRuntime.shared.deactivate()
        resolve(nil)
    }

    @objc(clear:rejecter:)
    func clear(
        _ resolve: RCTPromiseResolveBlock,
        rejecter reject: RCTPromiseRejectBlock
    ) {
        AnsightRuntime.shared.clear()
        resolve(nil)
    }

    @objc(metric:channel:resolver:rejecter:)
    func metric(
        _ value: String,
        channel: NSNumber?,
        resolver resolve: RCTPromiseResolveBlock,
        rejecter reject: RCTPromiseRejectBlock
    ) {
        do {
            try AnsightRuntime.shared.metric(Int64(value) ?? 0, channel: channel?.intValue ?? AnsightChannels.unspecified)
            resolve(nil)
        } catch {
            reject("ansight_metric_failed", error.localizedDescription, error)
        }
    }

    @objc(event:options:resolver:rejecter:)
    func event(
        _ label: String,
        options: NSDictionary?,
        resolver resolve: RCTPromiseResolveBlock,
        rejecter reject: RCTPromiseRejectBlock
    ) {
        do {
            let type = (options?["type"] as? String).flatMap(AnsightEventType.init(rawValue:)) ?? .info
            try AnsightRuntime.shared.event(
                label,
                type: type,
                details: options?["details"] as? String,
                channel: (options?["channel"] as? NSNumber)?.intValue ?? AnsightChannels.unspecified,
                id: options?["id"] as? String ?? UUID().uuidString
            )
            resolve(nil)
        } catch {
            reject("ansight_event_failed", error.localizedDescription, error)
        }
    }

    @objc(openSession:options:resolver:rejecter:)
    func openSession(
        _ pairingJson: String,
        options: NSDictionary,
        resolver resolve: RCTPromiseResolveBlock,
        rejecter reject: RCTPromiseRejectBlock
    ) {
        do {
            let result = try AnsightRuntime.shared.openSession(
                pairingJson: pairingJson,
                options: PairingOpenOptions(
                    clientName: options["clientName"] as? String ?? "",
                    expectedAppId: options["expectedAppId"] as? String,
                    profileOverride: options["profileOverride"] as? [String: String] ?? [:]
                )
            )

            resolve([
                "success": result.success,
                "message": result.message,
                "sessionId": result.sessionId as Any,
                "configId": result.configId as Any,
                "appId": result.appId as Any,
                "resolvedHostAddress": result.resolvedHostAddress as Any,
                "usedEmbeddedDeveloperPairing": result.usedEmbeddedDeveloperPairing,
                "discoverySource": result.discoverySource as Any,
            ])
        } catch {
            reject("ansight_open_session_failed", error.localizedDescription, error)
        }
    }

    @objc(completeSession:rejecter:)
    func completeSession(
        _ resolve: RCTPromiseResolveBlock,
        rejecter reject: RCTPromiseRejectBlock
    ) {
        AnsightRuntime.shared.completeSession()
        resolve(nil)
    }

    @objc(closeSession:rejecter:)
    func closeSession(
        _ resolve: RCTPromiseResolveBlock,
        rejecter reject: RCTPromiseRejectBlock
    ) {
        AnsightRuntime.shared.closeSession()
        resolve(nil)
    }

    @objc(registerTool:resolver:rejecter:)
    func registerTool(
        _ tool: NSDictionary,
        resolver resolve: RCTPromiseResolveBlock,
        rejecter reject: RCTPromiseRejectBlock
    ) {
        do {
            try AnsightRuntime.shared.registerTool(
                AnsightToolDescriptor(
                    id: tool["id"] as? String ?? "",
                    name: tool["name"] as? String ?? "",
                    scope: tool["scope"] as? String ?? "Read"
                )
            )
            resolve(nil)
        } catch {
            reject("ansight_register_tool_failed", error.localizedDescription, error)
        }
    }

    @objc(getDebugSnapshot:rejecter:)
    func getDebugSnapshot(
        _ resolve: RCTPromiseResolveBlock,
        rejecter reject: RCTPromiseRejectBlock
    ) {
        let snapshot = AnsightRuntime.shared.snapshot()
        resolve([
            "initialized": snapshot.initialized,
            "active": snapshot.active,
            "sessionOpen": snapshot.sessionOpen,
            "metricsRecorded": snapshot.metricsRecorded,
            "eventsRecorded": snapshot.eventsRecorded,
            "registeredTools": snapshot.registeredTools,
            "executableTools": snapshot.executableTools,
            "toolDiscoveryEnabled": snapshot.toolDiscoveryEnabled,
            "toolExecutionEnabled": snapshot.toolExecutionEnabled,
            "embeddedDeveloperPairingAvailable": snapshot.embeddedDeveloperPairingAvailable,
            "detectedBundledTools": snapshot.detectedBundledTools,
            "sessionMessage": snapshot.sessionMessage as Any,
            "lastPairingConfigId": snapshot.lastPairingConfigId as Any,
            "resolvedHostAddress": snapshot.resolvedHostAddress as Any,
            "lastMetric": snapshot.lastMetric.map {
                [
                    "value": $0.value,
                    "channel": $0.channel,
                    "capturedAtEpochMs": $0.capturedAtEpochMs,
                ]
            } as Any,
            "lastEvent": snapshot.lastEvent.map {
                [
                    "id": $0.id,
                    "label": $0.label,
                    "type": $0.type.rawValue,
                    "details": $0.details as Any,
                    "channel": $0.channel,
                    "capturedAtEpochMs": $0.capturedAtEpochMs,
                ]
            } as Any,
        ])
    }
}

private extension NSDictionary? {
    func toOptions() -> AnsightOptions {
        guard let options = self else {
            return AnsightOptions()
        }

        let channels = (options["additionalChannels"] as? [[String: Any]] ?? []).compactMap { raw in
            guard let id = raw["id"] as? NSNumber, let name = raw["name"] as? String else {
                return nil
            }

            return AnsightChannel(id: id.intValue, name: name, colorHex: raw["colorHex"] as? String)
        }

        return AnsightOptions(
            sampleFrequencyMilliseconds: (options["sampleFrequencyMilliseconds"] as? NSNumber)?.intValue ?? 500,
            retentionPeriodSeconds: (options["retentionPeriodSeconds"] as? NSNumber)?.intValue ?? 600,
            enableFramesPerSecond: (options["enableFramesPerSecond"] as? NSNumber)?.boolValue ?? true,
            additionalChannels: channels,
            toolGuard: (options["toolAccess"] as? String).map { rawValue in
                switch rawValue.lowercased() {
                case "readonly", "read":
                    return .readOnly
                case "all", "full":
                    return .fullAccess
                default:
                    return .disabled
                }
            } ?? .disabled
        )
    }
}
